src/WriteStreamJSON.h

```diff
@@ -0,0 +1,206 @@
+          beautify(getOr(rwOperation, "beautify", "false") == "ture"){};
+
+    const bool beautify;
+
+    void writeNextTupleJSON(std::ostream& destination, const RamDomain* tuple) {
+        std::vector<Json> result;
+
+        destination << "[";
+        for (size_t col = 0; col < arity; ++col) {
+            if (col > 0) {
+                destination << ", ";
+            }
+            writeNextTupleElement(destination, typeAttributes.at(col), tuple[col]);
+        }
+
+        // Output a JSON array for all tuples
+        destination << "]";
+    }
+
+    void writeNextTupleElement(std::ostream& destination, const std::string& name, const RamDomain value) {
+        using ValueTuple = std::pair<const std::string, const RamDomain>;
+        std::stack<std::variant<ValueTuple, std::string>> worklist;
+        worklist.push(std::make_pair(name, value));
+        
+        // the Json11 output is not tail recursive, therefore highly inefficient for recursive record
+        // in addition the JSON object is immutable, so has memory overhead
+        while(!worklist.empty()) {
+            std::variant<ValueTuple, std::string> curr = worklist.top();
+            worklist.pop();
+
+            if (std::holds_alternative<std::string>(curr)) {
+                destination << std::get<std::string>(curr);
+                continue;
+            }
+
+            const std::string& currType = std::get<ValueTuple>(curr).first;            
+            const RamDomain currValue = std::get<ValueTuple>(curr).second;
+            assert(currType.length() > 2 && "Invalid type length");
+            // the first two char is attribute char + ':', ignore them
+            // const char* currName = currType.c_str();
+            // destination << Json(currName + 2).dump().c_str() << ": ";
+            switch (currType[0]) {
+                // since some strings may need to be escaped, we use dump here
+                case 's': destination << Json(symbolTable.unsafeResolve(currValue)).dump(); break;
+                case 'i': destination << currValue; break;
+                case 'u': destination << (int) ramBitCast<RamUnsigned>(currValue); break;
+                case 'f': destination << ramBitCast<RamFloat>(currValue); break;
+                case 'r': {
+                    auto&& recordInfo = types["records"][currType];
+                    assert(!recordInfo.is_null() && "Missing record type information");
+                    if ((int) currValue == 0) {
+                        destination << "null";
+                        break;
+                    }
+
+                    auto&& recordTypes = recordInfo["types"];
+                    const size_t recordArity = recordInfo["arity"].long_value();
+                    const RamDomain* tuplePtr = recordTable.unpack(currValue, recordArity);
+                    worklist.push("]");              
+                    for (long long i = (long long) (recordArity - 1); i >= 0; --i) { 
```

Comment:
We use `size_t` for sizes/arities. This is a little inconvenient when counting down to 0, though. An alternative if that gets even less readable would be `int32_t`, with a single cast on (`recordArity`). That should safely contain all possible arities.